src/main/java/org/galatea/starter/service/IexService.java

```diff
@@ -59,10 +72,115 @@
     if (symbols.length() == 0) {
       return Collections.emptyList();
     } else if (date == null) {
-      return newClient.getHistoricalPricesWithRange(symbols, range);
+      checkAndRetrieveForRange(service, symbols, range);
+      return retrieveFromDBForRange(service, symbols, range);
     } else {
-      return newClient.getHistoricalPricesWithDate(symbols, date);
+      checkAndRetrieveForDate(service, symbols, date);
+      return retrieveFromDBForDate(service, symbols, date);
+    }
+  }
+
+  public void checkAndRetrieveForRange(HistoricalPriceDBService service, String symbol,
+      String rangeOfDays) {
+    String r = rangeOfDays.split("(?<=\\D)(?=\\d)|(?<=\\d)(?=\\D)")[0];
+    int range = 30;
+    if (r.length() != 0) {
+      range = Integer.parseInt(r);
+    }
+    LocalDate today = LocalDate.now();
+    for (int i = range; i >= 1; i--) {
+      LocalDate date = today.minusDays(i);
+      DateTimeFormatter formatters = DateTimeFormatter.ofPattern("yyyyMMdd");
+      CompositePrimaryKey obj = new CompositePrimaryKey(symbol, date,
+          LocalTime.now().withHour(0).withMinute(0).withSecond(0).withNano(0));
+      if (!service.exists(obj)) {
+        List<IexHistoricalPrice> prices =
+            newClient.getHistoricalPriceWithDateByDay(symbol, date.format(formatters));
+        for (IexHistoricalPrice price : prices) {
+          LocalTime t = LocalTime.now().withHour(0).withMinute(0).withSecond(0).withNano(0);
+          HistoricalPriceDB objToSave =
+              new HistoricalPriceDB(price.getClose(), price.getHigh(), price.getLow(),
+                  price.getOpen(), price.getSymbol(), price.getVolume(), price.getDate(), t,
+                  LocalTime.now());
+          service.save(objToSave);
+        }
+      }
+    }
+  }
+
+  public List<IexHistoricalPrice> retrieveFromDBForRange(HistoricalPriceDBService service,
+      String symbol,
+      String rangeOfDays) {
+    List<IexHistoricalPrice> returnList = new ArrayList<>();
+    String r = rangeOfDays.split("(?<=\\D)(?=\\d)|(?<=\\d)(?=\\D)")[0];
```

Comment:
this is the same logic as the above method. can we combine functionality?

Reply:
This should be resolved now.